noir_stdlib/src/hash/sha256.nr

```diff
@@ -169,99 +169,352 @@ unconstrained fn build_msg_block<let N: u32>(
+//     update_block_item(
+//         msg_block,
+//         msg_byte_ptr,
+//         |msg_item| set_item_byte(msg_item, msg_byte_ptr, msg_byte),
+//     )
+// }
+
+// // Replace one byte in a message block item.
+// fn set_item_byte(msg_item: u32, msg_byte_ptr: BLOCK_BYTE_PTR, msg_byte: u8) -> u32 {
+//     let old_byte = get_item_byte(msg_item, msg_byte_ptr);
+//     let old_item = byte_into_item(old_byte, msg_byte_ptr);
+//     let new_item = byte_into_item(msg_byte, msg_byte_ptr);
+//     msg_item - old_item + new_item
+// }
+
+// Set the rightmost `zeros` number of bytes to 0.
+fn set_item_zeros(item: u32, zeros: u8) -> u32 {
+    lshift8(rshift8(item, zeros), zeros)
+}
+
+// Replace one byte in the item with a value, and set everything after it to zero.
+fn set_item_byte_then_zeros(msg_item: u32, msg_byte_ptr: BLOCK_BYTE_PTR, msg_byte: u8) -> u32 {
+    let zeros = INT_SIZE - msg_byte_ptr % INT_SIZE;
+    let zeroed_item = set_item_zeros(msg_item, zeros as u8);
+    let new_item = byte_into_item(msg_byte, msg_byte_ptr);
+    zeroed_item + new_item
+}
+
+// Get a byte of a message item according to its overall position in the `BLOCK_SIZE` space.
+fn get_item_byte(mut msg_item: u32, msg_byte_ptr: BLOCK_BYTE_PTR) -> u8 {
+    // How many times do we have to shift to the right to get to the position we want?
+    let max_shifts = INT_SIZE - 1;
+    let shifts = max_shifts - msg_byte_ptr % INT_SIZE;
+    msg_item = rshift8(msg_item, shifts as u8);
+    // At this point the byte we want is in the rightmost position.
+    msg_item as u8
+}
+
+// Project a byte into a position in a field based on the overall block pointer.
+// For example putting 1 into pointer 5 would be 100, because overall we would
+// have [____, 0100] with indexes [0123,4567].
+fn byte_into_item(msg_byte: u8, msg_byte_ptr: BLOCK_BYTE_PTR) -> u32 {
+    let mut msg_item = msg_byte as u32;
+    // How many times do we have to shift to the left to get to the position we want?
+    let max_shifts = INT_SIZE - 1;
+    let shifts = max_shifts - msg_byte_ptr % INT_SIZE;
+    lshift8(msg_item, shifts as u8)
+}
+
+// Construct a field out of 4 bytes.
+fn make_item(b0: u8, b1: u8, b2: u8, b3: u8) -> u32 {
+    let mut item = b0 as u32;
+    item = lshift8(item, 1) + b1 as u32;
+    item = lshift8(item, 1) + b2 as u32;
+    item = lshift8(item, 1) + b3 as u32;
+    item
+}
+
+// Shift by 8 bits to the left between 0 and 4 times.
+// Checks `is_unconstrained()` to just use a bitshift if we're running in an unconstrained context,
+// otherwise multiplies by 256.
+fn lshift8(item: u32, shifts: u8) -> u32 {
+    if is_unconstrained() {
+        if item == 0 {
+            0
+        } else {
+            // Brillig wouldn't shift 0<<4 without overflow.
+            item << (8 * shifts)
+        }
+    } else {
+        // We can do a for loop up to INT_SIZE or an if-else.
+        if shifts == 0 {
+            item
+        } else if shifts == 1 {
+            item * 256
```

Comment:
Did you notice a difference with just using `<<`? The compiler should simplify `<<` to `*`

Reply:
I didn't try, just kept this because of how it was previously. Tom recommended to use bit shifting in unconstrained because it can deal with any input, whereas in the circuit it has to be known. I'll try, maybe it can be used to avoid having globals naming all these.

Reply:
Ah I see it is because we do not necessarily know the `shifts` at compile time.

Reply:
Tried it with bit shifting and the globals `TWO_POW_...`s. The latter is 10% faster.